corehq/apps/case_search/xpath_functions/ancestor_functions.py

```diff
@@ -92,22 +91,73 @@ def _is_ancestor_path_expression(node):
-        new_query = '{} {} "{}"'.format(serialize(node.left.right), node.op, node.right)
-        raise TooManyRelatedCasesError(
-            gettext("The related case lookup you are trying to perform would return too many cases"),
-            new_query
-        )
+    return CaseSearchES().domain(context.domain).get_child_cases(case_ids, identifier).scroll_ids()
 
-    return es_query.scroll_ids()
 
+def ancestor_exists(node, context):
+    """
+    Supports the following syntax:
+    - ancestor_exists({ancestor_path_expression}, {ancestor_filter})
+    i.e. ancestor-exists(host/parent,city='SF')
 
-def _child_case_lookup(context, case_ids, identifier):
-    """returns a list of all case_ids who have parents `case_id` with the relationship `identifier`
+    Since ES has no way of performing joins, we filter down in stages:
+    1. Find the ids of all cases where the condition is met
+    2. Walk down the case hierarchy, finding all related cases with the right identifier to the ids
+    found in (1).
+    3. Return the lowest of these ids as a related case query filter
     """
-    return CaseSearchES().domain(context.domain).get_child_cases(case_ids, identifier).scroll_ids()
+    confirm_args_count(node, 2)
+    ancestor_path_node, ancestor_case_filter_node = node.args
+    return process_ancestor_exists(ancestor_path_node, ancestor_case_filter_node, context)
+
+
+def process_ancestor_exists(ancestor_path_node, ancestor_case_filter_node, context):
+    _validate_ancestor_exists_filter(ancestor_case_filter_node)
+    base_case_ids = _get_case_ids_from_ast_filter(context, ancestor_case_filter_node)
+
+    return walk_ancestor_hierarchy(context, ancestor_path_node, base_case_ids)
+
+
+def _validate_ancestor_exists_filter(node):
+    """
+    ancestor_exists(status="active" and subcase-exists(...)) raises error
+    """
+
+    incompatible_xpath_query_functions = {'subcase-exists', 'subcase-count'}
+
+    if isinstance(node, FunctionCall) and node.name in incompatible_xpath_query_functions:
+        raise CaseFilterError(gettext(f"{node.name} is not supported with ancestor-exists"), node.name)
+
+    if hasattr(node, 'op') and node.op in OPERATOR_MAPPING:
+        # This node is a leaf
+        return OPERATOR_MAPPING[node.op](_validate_ancestor_exists_filter(node.left),
+                                        _validate_ancestor_exists_filter(node.right))
+    return
+
+
+def _get_case_ids_from_ast_filter(context, filter_node):
+    from corehq.apps.case_search.dsl_utils import unwrap_value
+    if (isinstance(filter_node, BinaryExpression)
+    and serialize(filter_node.left) == "@case_id" and filter_node.op in (EQ, NEQ)):
+        # case id is provided in query i.e @case_id="b9eaf791-e427-482d-add4-2a60acf0362e"
+        case_ids = unwrap_value(filter_node.right, context)
+        if isinstance(case_ids, str):
+            return (case_id for case_id in [case_ids])
+        else:
+            return (case_id for case_id in case_ids)
```

Comment:
I don't think it's necessary to make these generators:
```suggestion
        return [case_ids] if isinstance(case_ids, str) else case_ids
```

Reply:
I don't think there would need to be any type checking as long as the caller treats the returned value as an iterable.